src/mono/wasm/runtime/storage.ts

```diff
@@ -0,0 +1,170 @@
+        const cache = await openCache();
+        if (!cache) {
+            return undefined;
+        }
+        const res = await cache.match(cacheKey);
+        if (!res) {
+            return undefined;
+        }
+        return res.arrayBuffer();
+    } catch (ex) {
+        console.warn("MONO_WASM: Failed load memory snapshot from the cache", ex);
+        return undefined;
+    }
+}
+
+export async function storeMemorySnapshot(memory: ArrayBuffer) {
+    try {
+        const inputsHash = await getInputsHash();
+        if (!inputsHash) {
+            return;
+        }
+        const cacheKey = `${memoryPrefix}-${ProductVersion}-${GitHash}-${inputsHash}`;
+        const cache = await openCache();
+        if (!cache) {
+            return;
+        }
+        const copy = MonoWasmThreads
+            // storing SHaredArrayBuffer in the cache is not working
+            ? (new Int8Array(memory)).slice(0)
+            : memory;
+
+        const responseToCache = new Response(copy, {
+            headers: {
+                "content-type": "wasm-memory",
+                "content-length": memory.byteLength.toString(),
+            },
+        });
+
+        await cache.put(cacheKey, responseToCache);
+
+        cleanupMemorySnapshots(cacheKey); // no await
+    } catch (ex) {
+        console.warn("MONO_WASM: Failed to store memory snapshot in the cache", ex);
+        return;
+    }
+}
+
+export async function cleanupMemorySnapshots(protectKey: string) {
+    try {
+        const cache = await openCache();
+        if (!cache) {
+            return;
+        }
+        const items = await cache.keys();
+        for (const item of items) {
+            if (item.url && item.url !== protectKey && item.url.startsWith(memoryPrefix)) {
+                await cache.delete(item);
+            }
+        }
+    } catch (ex) {
+        return;
+    }
+}
+
+// calculate hash of things which affect the memory snapshot
+export async function getInputsHash(): Promise<string | null> {
+    if (!runtimeHelpers.subtle) {
+        return null;
+    }
+    const inputs = Object.assign({}, runtimeHelpers.config) as any;
+    // above already has env variables, runtime options, etc
+    // above also already has config.assetsHash for this. It has all the asserts (DLLs, ICU, .wasms, etc). 
+    // So we could remove assets collectionfrom the hash.
+    delete inputs.assets;
+    // some things are calculated at runtime, so we need to add them to the hash
+    inputs.preferredIcuAsset = runtimeHelpers.preferredIcuAsset;
+    inputs.timezone = runtimeHelpers.timezone;
+    // some things are not relevant for memory snapshot
+    delete inputs.forwardConsoleLogsToWS;
+    delete inputs.diagnosticTracing;
+    delete inputs.appendElementOnExit;
+    delete inputs.logExitCode;
+    delete inputs.pthreadPoolSize;
+    delete inputs.asyncFlushOnExit;
+    delete inputs.assemblyRootFolder;
+    delete inputs.remoteSources;
```

Comment:
Not your problem to fix necessarily, but SHA integrity is probably compromised if you ever loaded anything from a remote source, since the remote source could deploy a new version of it at any time.
We might want to consider disabling snapshots if remote sources are in use at all. (Do customers use them? The intent was to allow hosting ICU data on a CDN, but I don't know if anyone ever did it.)

Reply:
`remoteSources` needs to comply with the `integrity` of the asset. 
We are adding them to `fetch` request already. 
Otherwise the download itself would fail.
That probably means that people would need to have namespaces in their CDN url.